### PR TITLE
feat: List tracks in observations list

### PR DESCRIPTION
**Pull request by ErikSin**
Adds recorded tracks to observations list. **Clicking on the track will not do anything, that is being handled in a follow up PR**

image

### CODE REVIEW

File: src/frontend/screens/MapScreen/ObsevationMapLayer.tsx [hunk 1]
```diff
@@ -15,7 +15,7 @@ const layerStyles = {
 };
 
 export const ObservationMapLayer = () => {
-  const observations = useAllObservations();
+  const {data: observations} = useObservations();
```
Comment (ErikSin) [inline src/frontend/screens/MapScreen/ObsevationMapLayer.tsx:18]: Right now the user only see the observations on the home screen map, but I have asked Sabella if they should also see the tracks.

**Review by achou11 — CHANGES_REQUESTED** [review]
mostly minor comments but a few that are related to potentially accounting for the feature flag

File: src/frontend/screens/MapScreen/ObsevationMapLayer.tsx (hunk 1 above)
Comment (ErikSin) [reply src/frontend/screens/MapScreen/ObsevationMapLayer.tsx:18]: Added tracks as a map layer: https://github.com/digidem/comapeo-mobile/pull/408/commits/32245c48431a56ec7050794476a2acf647b22ba4

File: src/frontend/hooks/useObservationsAndTracks.ts
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/hooks/useObservationsAndTracks.ts:1]: yeah, I agree, having it as a seperate hook feels a little unnecessary. So I just got rid of this hook: https://github.com/digidem/comapeo-mobile/pull/408/commits/48853ee0d5ea7185069d27624837dba0b09917ec

File: src/frontend/screens/ObservationsList/index.tsx [hunk 2]
```diff
@@ -84,19 +90,33 @@ export const ObservationsList: React.FC<
...
-              }
-            />
-          );
+          switch (item.schemaName) {
+            case 'observation':
+              return (
+                <ObservationListItem
+                  key={item.docId}
+                  testID={`observationListItem:${index}`}
+                  observation={item as Observation}
```
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/screens/ObservationsList/index.tsx]: https://github.com/digidem/comapeo-mobile/pull/408/commits/53304d45ecb2837fad195016bb3496e3ec305149

File: src/frontend/screens/ObservationsList/index.tsx [hunk 3]
```diff
@@ -84,19 +90,33 @@ export const ObservationsList: React.FC<
...
+                      observationId: item.docId,
+                    })
+                  }
+                />
+              );
+            case 'track':
+              return (
+                <TrackListItem
+                  testID={`trackListItem:${index}`}
+                  track={item as Track}
```
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/screens/ObservationsList/index.tsx]: https://github.com/digidem/comapeo-mobile/pull/408/commits/53304d45ecb2837fad195016bb3496e3ec305149

File: src/frontend/screens/MapScreen/ObsevationMapLayer.tsx
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/screens/MapScreen/ObsevationMapLayer.tsx:1]: https://github.com/digidem/comapeo-mobile/pull/408/commits/48853ee0d5ea7185069d27624837dba0b09917ec

File: src/frontend/screens/ObservationsList/TrackListItem.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        </View>
+        <TrackIcon />
+      </View>
+    </TouchableHighlight>
+  );
+};
+
+export const TrackListItem = React.memo<ObservationListItemProps>(
+  TrackObservationItemNotMemoized,
+);
```
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/screens/ObservationsList/TrackListItem.tsx:55]: I think they designed this based off `ObservationListItem` which was memoized.

I don't think its a bad thing to use `React.memo`. There is minimal overhead (ie. no dependency array like useMemo) and will optimize component

File: src/frontend/screens/ObservationsList/index.tsx [hunk 5]
```diff
@@ -84,19 +90,33 @@ export const ObservationsList: React.FC<
...
+                  style={styles.listItem}
+                  onPress={() => {}}
```
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/screens/ObservationsList/index.tsx]: https://github.com/digidem/comapeo-mobile/pull/408/commits/c71b29082a9a584e2acd2cced32265d93c0b55ad

File: src/frontend/screens/ObservationsList/TrackListItem.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  track,
+  testID,
+  onPress,
+}: ObservationListItemProps) => {
+  const {formatMessage} = useIntl();
+  return (
+    <TouchableHighlight
+      onPress={onPress}
+      testID={testID}
+      style={{flex: 1, height: 80}}>
```
[the remark this answers is not in the document]
Comment (ErikSin) [reply src/frontend/screens/ObservationsList/TrackListItem.tsx]: https://github.com/digidem/comapeo-mobile/pull/408/commits/69de7eb54c05e979c056eab31868539013c711c7

File: src/frontend/hooks/server/track.ts [hunk 7]
```diff
@@ -23,13 +23,15 @@ export function useCreateTrack() {
   });
 }
 
-export function useTracksQuery() {
+export function useTracks() {
```
Comment (ErikSin) [inline src/frontend/hooks/server/track.ts:26]: I renamed this to be consistent with the rest of the server hooks

File: src/frontend/screens/MapScreen/CurrentTrack/CurrrentTrackMapLayer.tsx [hunk 8]
```diff
@@ -6,7 +6,7 @@ import {LineString} from 'geojson';
 import {useLocation} from '../../../hooks/useLocation';
 import {LocationHistoryPoint} from '../../../sharedTypes/location';
 
-export const TrackPathLayer = () => {
+export const CurrentTrackMapLayer = () => {
```
Comment (ErikSin) [inline src/frontend/screens/MapScreen/CurrentTrack/CurrrentTrackMapLayer.tsx:9]: I renamed this so there was a distinction between the track being made and previous tracks shown on the map

File: src/frontend/screens/MapScreen/index.tsx [hunk 9]
```diff
@@ -104,8 +105,13 @@ export const MapScreen = () => {
         {coords && locationServicesEnabled && (
           <UserLocation minDisplacement={MIN_DISPLACEMENT} />
         )}
-        {isFinishedLoading && <ObservationMapLayer />}
-        {isFinishedLoading && <TrackPathLayer />}
+        {isFinishedLoading && (
+          <>
+            <ObservationMapLayer />
+            <TracksMapLayer />
+            <CurrentTrackMapLayer />
```
Comment (achou11) [inline src/frontend/screens/MapScreen/index.tsx:112]: should these be flipped? my understanding is that the existing tracks will cover the current track if implemented like this

File: src/frontend/screens/ObservationsList/TrackListItem.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (achou11) [reply src/frontend/screens/ObservationsList/TrackListItem.tsx:55]: > There is minimal overhead (ie. no dependency array like useMemo) and will optimize component

i think using [memo](https://react.dev/reference/react/memo#memo) like this is similar to adding _every prop_ to a dependency array (i.e. checking each one as they change), so there's technically more overhead in this case. but yeah, agree that it's probably not a huge issue in this case

File: src/frontend/screens/ObservationsList/index.tsx [hunk 10]
```diff
@@ -84,19 +89,37 @@ export const ObservationsList: React.FC<
...
+            case 'track':
+              return (
+                <TrackListItem
+                  testID={`trackListItem:${index}`}
+                  track={item}
+                  style={styles.listItem}
+                  onPress={() => {
+                    // TODO: navigate to track
+                  }}
+                />
```
Comment (achou11) [inline src/frontend/screens/ObservationsList/index.tsx:116]: add a `key`

File: src/frontend/screens/MapScreen/index.tsx (hunk 9 above)
Comment (achou11) [reply src/frontend/screens/MapScreen/index.tsx:112]: also, might be worth hiding behind the feature flag here

**Review by achou11 — APPROVED** [review]
few more comments that should be addressed but not blocking

**Comment by ErikSin** [comment]
> i think using [memo](https://react.dev/reference/react/memo#memo) like this is similar to adding every prop to a dependency array (i.e. checking each one as they change), so there's technically more overhead in this case. but yeah, agree that it's probably not a huge issue in this case

Just to clarify. I more so meant that there is no overhead for developers. My main reasoning for not memoizing everything in a react function is because it becomes hard to maintain as complexity increases. Dependency arrays become long, and the code just becomes quite difficult to read. That doesn't really apply to React.memo as all of that is internally kept track off, so I generally don't feel the need to not use it.

File: src/frontend/screens/ObservationsList/index.tsx (hunk 10 above)
Comment (ErikSin) [reply src/frontend/screens/ObservationsList/index.tsx]: https://github.com/digidem/comapeo-mobile/pull/408/commits/4eee0ac214ab66a52c294b5691b2fc1e6c767566

File: src/frontend/screens/MapScreen/index.tsx (hunk 9 above)
Comment (ErikSin) [reply src/frontend/screens/MapScreen/index.tsx]: https://github.com/digidem/comapeo-mobile/pull/408/commits/e66d44c1a0bf44df03042cdcf7c45ff8ec1d3331

**Pull request merged**